README.md
node-solr-smart-client
======================

A node.js solr smart client. In short it serves as a smart constructor for [solr-client](https://github.com/lbdremy/solr-node-client). The constructor queries a ZooKeeper
ensemble for live nodes of the input Solr collection. When more than one live node is returned by ZooKeeper, the constructor picks one at random and hands over to solr-client.

Features
--------

* a smart constructor for [solr-client](https://github.com/lbdremy/solr-node-client)
* retrieval of live Solr nodes from ZooKeeper cluster using [node-zookeeper-client](https://github.com/alexguan/node-zookeeper-client)
* HTTP requests towards Solr nodes using [node-rest-client](https://github.com/aacerox/node-rest-client)
* supports SSH tunneling
* supports Solr collection aliases

Installation
------------

    npm install node-solr-smart-client
    
Include option `--save` to add it to your package.json file in the same go:

    npm install node-solr-smart-client --save
    
Usage
-----

### Basic

    // Load dependency
    var solrSmartClient = require('node-solr-smart-client');
    
    // Define options
    options = {
        zkConnectionString: 'localhost:2181',
        zkLiveNodes: '/live_nodes',
        zkAliases: '/aliases.json',
        solrProtocol: 'http',
        solrCollectionsGetEndPoint: '/admin/collections?action=LIST&wt=json', // Supports XML and JSON writer types
        ssh: {},
        // Passed verbatim to node-zookeeper-client
        zk: {
            sessionTimeout: 3000,
            spinDelay : 1000,
            retries : 1
        },
        // Passed verbatim to node-rest-client
        rest: {
            requestConfig: {
                timeout: 3000
            },
            responseConfig: {
                timeout: 3000
            },
            mimetypes: {
                json: ["application/json", "application/json;charset=utf-8", "application/json; charset=utf-8", "application/json;charset=UTF-8", "application/json; charset=UTF-8"],
                xml: ["application/xml", "application/xml;charset=utf-8", "application/xml; charset=utf-8", "application/xml;charset=UTF-8", "application/xml; charset=UTF-8"]
            }
        }
    };
    
    // Create Solr client, execute query and print number of documents in response.
    solrSmartClient.createClient('my_solr_collection', options, function (err, solrClient) {
        if (err) {
            return console.log(err);
        }
        solrClient.search('q=*:*', function (err, obj) {
            if (err) {
                return console.log(err);
            }
            console.log('Number of documents found: %d', obj.response.numFound);
        })
    });
    
### SSH tunneling

Assuming SSH tunnels have been set up in the following manner

    ssh -f -N -L 2181:my_zookeeper_node_1:2181 my_user@my_zookeeper_node_1
    ssh -f -N -L 2182:my_zookeeper_node_2:2181 my_user@my_zookeeper_node_2
    ssh -f -N -L 8080:my_solr_node_1:8080 my_user@my_solr_node_1
    ssh -f -N -L 8081:my_solr_node_2:8080 my_user@my_solr_node_2
    
The options' `ssh` field may be set as

    ssh: {
        tunnels: {
            'my_solr_node_1:8080': 'localhost:8080',
            'my_solr_node_2:8080': 'localhost:8081'
        }
    }

or alternatively in SSH style as

    ssh: {
        tunnels: '8080:my_solr_node_1:8080,8081:my_solr_node_2:8080'
    }
    
As well `zkConnectionString` in `options` must have the tunneled value of `'localhost:2181,localhost:2182'`.
    
Test
----

    npm test


CHANGELOG.md
# Change Log

## 0.2.1

1. Fixed bug on propagation of `options.rest` to node-rest-client instance
2. Added default MIME types for XML and JSON to node-rest-client in support of Solr 5
3. Added default writer type `JSON` to end point retrieving Solr collections from API

## 0.2.0

1. Added support for Solr collection aliases
2. Revamped dependencies to most up-to-date minors
3. [Pull #1: Collections list from Solr as JSON](https://github.com/jijordre/node-solr-smart-client/pull/1)
4. Added this changelog and removed revision history from README.md

## 0.1.0

Initial release.


test/createClientTest.js
var should = require('should'),
    sinon = require('sinon'),
    proxyquire = require('proxyquire'),
    zkUtils = mockZkUtils(),
    restClientObj = {},
    restClient = {
        Client: function (options) {
            return restClientObj;
        }
    },
    solrClient = mockSolrClient();

var solrSmartClient = proxyquire('../lib/index.js', {
    'node-rest-client': restClient,
    './zookeeper-utils.js': zkUtils,
    'solr-client': solrClient
});

function mockZkUtils() {
    var zkUtils = {
        getChildren: sinon.stub(),
        getData: sinon.stub()
    };
    var some_live_nodes = ['some_solr_node1:1234_some_path', 'some_solr_node2:1234_some_path'];
    var empty_live_nodes = [];
    zkUtils.getChildren.withArgs(sinon.match.object, '/some_live_nodes', sinon.match.func).yields(null, some_live_nodes);
    zkUtils.getChildren.withArgs(sinon.match.object, '/empty_live_nodes', sinon.match.func).yields(null, empty_live_nodes);
    zkUtils.getChildren.withArgs(sinon.match.object, '/bogus_live_nodes', sinon.match.func).yields('some_error', null);
    zkUtils.getChildren.withArgs(sinon.match.object, null, sinon.match.func).yields('some_error', null);
    var some_aliases = new Buffer(JSON.stringify({'collection': {'some_collection_alias': 'some_collection2'}}));
    var empty_aliases = new Buffer(JSON.stringify({}));
    zkUtils.getData.withArgs(sinon.match.object, '/some_aliases.json', sinon.match.func).yields(null, some_aliases);
    zkUtils.getData.withArgs(sinon.match.object, '/empty_aliases.json', sinon.match.func).yields(null, empty_aliases);
    zkUtils.getData.withArgs(sinon.match.object, '/bogus_aliases.json', sinon.match.func).yields('some_error', null);
    zkUtils.getData.withArgs(sinon.match.object, null, sinon.match.func).yields('some_error', null);
    return zkUtils;
}

function mockRESTClientObj() {
    var xmlData = {
        response: {
            arr: [
                {
                    str: ['some_collection1', 'some_collection2', 'some_collection3']
                }
            ]
        }
    };
    var xmlResponse = {
        headers: {
            'content-type': 'application/xml;charset=UTF-8'
        }
    };
    var jsonData = {
        collections: ['some_collection1', 'some_collection2', 'some_collection3']
    };
    var jsonResponse = {
        headers: {
            'content-type': 'application/json;charset=UTF-8'
        }
    };
    var unsupportedResponse = {
        headers: {
            'content-type': 'text/plain;charset=UTF-8'
        },
        req: {
            path: '/some/request/path'
        }
    }
    restClientObj.get = sinon.stub();
    restClientObj.get.withArgs(sinon.match(/wt=xml/), sinon.match.object, sinon.match.func).yields(xmlData, xmlResponse);
    restClientObj.get.withArgs(sinon.match(/wt=json/), sinon.match.object, sinon.match.func).yields(jsonData, jsonResponse);
    restClientObj.get.withArgs(sinon.match(/wt=(?!xml|json)/), sinon.match.object, sinon.match.func).yields(null, unsupportedResponse);
    restClientObj.get.withArgs(sinon.match(/^((?!wt=).)*$/), sinon.match.object, sinon.match.func).yields(xmlData, xmlResponse);
}

function mockSolrClient() {
    return {
        createClient: function (host, port, core, path) {
            return {
                options: {
                    host: host,
                    port: port,
                    core: core,
                    path: path
                }
            }
        }
    }
}

function mockOptions() {
    return {
        zkConnectionString: 'some_zookeeper_host:0000',
        zkLiveNodes: '/some_live_nodes',
        zkAliases: '/some_aliases.json',
        solrProtocol: 'http',
        solrCollectionsGetEndPoint: '/admin/collections?action=LIST',
        ssh: {},
        // Passed verbatim to node-zookeeper-client
        zk: {
            sessionTimeout: 3000,
            spinDelay: 1000,
            retries: 1
        },
        // Passed verbatim to node-rest-client
        rest: {
            requestConfig: {
                timeout: 3000
            },
            responseConfig: {
                timeout: 3000
            },
            mimetypes: {
                json: ["application/json", "application/json;charset=utf-8", "application/json; charset=utf-8", "application/json;charset=UTF-8", "application/json; charset=UTF-8"],
                xml: ["application/xml", "application/xml;charset=utf-8", "application/xml; charset=utf-8", "application/xml;charset=UTF-8", "application/xml; charset=UTF-8"]
            }
        }
    };
}

describe('solrSmartClient', function () {

    var options;

    before(function() {
        sinon.spy(restClient, "Client");
    });

    beforeEach(function () {
        mockRESTClientObj();
        options = mockOptions();
    });

    after(function() {
       restClient.Client.restore();
    });

    describe('#createClient', function () {

        describe('with bogus ZooKeeper live nodes path', function () {

            beforeEach(function () {
                options.zkLiveNodes = '/bogus_live_nodes';
            });

            it('should throw error', function (done) {
                solrSmartClient.createClient('some_collection2', options, function (err, solrClient) {
                    should.exist(err);
                    should.not.exist(solrClient);
                    done();
                });
            });
        });

        describe('with no ZooKeeper live nodes path', function () {

            beforeEach(function () {
                options.zkLiveNodes = null;
            });

            it('should throw error', function (done) {
                solrSmartClient.createClient('some_collection2', options, function (err, solrClient) {
                    should.exist(err);
                    should.not.exist(solrClient);
                    done();
                });
            });
        });

        describe('with empty live nodes', function () {

            beforeEach(function () {
                options.zkLiveNodes = '/empty_live_nodes';
            });

            it('should throw error', function (done) {
                solrSmartClient.createClient('some_collection2', options, function (err, solrClient) {
                    should.exist(err);
                    should.not.exist(solrClient);
                    done();
                });
            });
        });

        describe('with no live node matching the name of the Solr collection', function () {

            it('should throw error', function (done) {
                solrSmartClient.createClient('no_collection', options, function (err, solrClient) {
                    should.exist(err);
                    should.not.exist(solrClient);
                    done();
                });
            });
        });

        describe('with target Solr collection being alias', function () {

            describe('with no ZooKeeper aliases path', function () {

                beforeEach(function () {
                    options.zkAliases = null;
                });

                it('should throw error', function (done) {
                    solrSmartClient.createClient('some_collection_alias', options, function (err, solrClient) {
                        should.exist(err);
                        should.not.exist(solrClient);
                        done();
                    });
                });
            });

            describe('with bogus ZooKeeper aliases path', function () {

                beforeEach(function () {
                    options.zkAliases = '/bogus_aliases.json';
                });

                it('should throw error', function (done) {
                    solrSmartClient.createClient('some_collection_alias', options, function (err, solrClient) {
                        should.exist(err);
                        should.not.exist(solrClient);
                        done();
                    });
                });
            });

            describe('with empty aliases', function () {

                beforeEach(function () {
                    options.zkAliases = '/empty_aliases.json';
                });

                it('should throw error', function (done) {
                    solrSmartClient.createClient('some_collection_alias', options, function (err, solrClient) {
                        should.exist(err);
                        should.not.exist(solrClient);
                        done();
                    });
                });
            });

            describe('with well-defined aliases', function () {

                it('should create instance of node-solr-client on alias', function (done) {
                    solrSmartClient.createClient('some_collection_alias', options, function (err, solrClient) {
                        should.not.exist(err);
                        should.exist(solrClient);
                        solrClient.options.port.should.eql('1234');
                        solrClient.options.path.should.eql('/some/path/some_collection_alias');
                        done();
                    });
                });
            });
        });

        describe('with JSON writer type', function() {

            beforeEach(function() {
                options.solrCollectionsGetEndPoint = '/admin/collections?action=LIST&wt=json';
            });

            it('should create instance of node-solr-client', function (done) {
                solrSmartClient.createClient('some_collection2', options, function (err, solrClient) {
                    should.not.exist(err);
                    should.exist(solrClient);
                    solrClient.options.host.should.startWith('some_solr_node');
                    solrClient.options.port.should.eql('1234');
                    solrClient.options.path.should.eql('/some/path/some_collection2');
                    done();
                });
            });
        })

        describe('with XML writer type', function () {

            beforeEach(function() {
                options.solrCollectionsGetEndPoint = '/admin/collections?action=LIST&wt=xml';
            });

            it('should create instance of node-solr-client', function (done) {
                solrSmartClient.createClient('some_collection2', options, function (err, solrClient) {
                    should.not.exist(err);
                    should.exist(solrClient);
                    solrClient.options.host.should.startWith('some_solr_node');
                    solrClient.options.port.should.eql('1234');
                    solrClient.options.path.should.eql('/some/path/some_collection2');
                    done();
                });
            });
        });

        describe('with default writer type', function () {

            it('should create instance of node-solr-client', function (done) {
                solrSmartClient.createClient('some_collection2', options, function (err, solrClient) {
                    should.not.exist(err);
                    should.exist(solrClient);
                    solrClient.options.host.should.startWith('some_solr_node');
                    solrClient.options.port.should.eql('1234');
                    solrClient.options.path.should.eql('/some/path/some_collection2');
                    done();
                });
            });
        });

        describe('with unsupported writer type', function () {

            beforeEach(function() {
                options.solrCollectionsGetEndPoint = '/admin/collections?action=LIST&wt=csv';
            });

            it('should throw error', function (done) {
                solrSmartClient.createClient('some_collection2', options, function (err, solrClient) {
                    should.exist(err);
                    should.not.exist(solrClient);
                    done();
                });
            });
        });

        describe('with SSH tunnel specification', function () {

            describe('when specified as object', function () {

                beforeEach(function () {
                    options.ssh = {
                        tunnels: {
                            'some_solr_node1:1234': 'localhost:4321',
                            'some_solr_node2:1234': 'localhost:4321'
                        }
                    };
                });

                it('should tunnel requests to Solr servers', function (done) {
                    solrSmartClient.createClient('some_collection2', options, function (err, solrClient) {
                        solrClient.options.host.should.startWith('localhost');
                        solrClient.options.port.should.eql('4321');
                    });
                    done();
                });
            });

            describe('when specified as string', function () {

                beforeEach(function () {
                    options.ssh = {
                        tunnels: '4321:some_solr_node1:1234,4321:some_solr_node2:1234'
                    };
                });

                it('should parse tunnel configs fed as string and tunnel requests to Solr servers', function (done) {
                    solrSmartClient.createClient('some_collection2', options, function (err, solrClient) {
                        solrClient.options.host.should.startWith('localhost');
                        solrClient.options.port.should.eql('4321');
                    });
                    done();
                });
            });
        });

        describe('with REST client options specification', function() {

            it('should instantiate REST client with options.rest', function(done) {
                solrSmartClient.createClient('some_collection2', options, function (err, solrClient) {
                    restClient.Client.calledWith(options.rest);
                    done();
                });
            });
        });
    });
});


examples/query.js
// Load dependency
var solrSmartClient = require('../lib/index.js');

// Define options
options = {
    zkConnectionString: 'localhost:2181',
    zkLiveNodes: '/live_nodes',
    zkAliases: '/aliases.json',
    solrProtocol: 'http',
    solrCollectionsGetEndPoint: '/admin/collections?action=LIST&wt=json', // Supports XML and JSON writer types
    ssh: {},
    // Passed verbatim to node-zookeeper-client
    zk: {
        sessionTimeout: 3000,
        spinDelay : 1000,
        retries : 1
    },
    // Passed verbatim to node-rest-client
    rest: {
        requestConfig: {
            timeout: 3000
        },
        responseConfig: {
            timeout: 3000
        },
        mimetypes: {
            json: ["application/json", "application/json;charset=utf-8", "application/json; charset=utf-8", "application/json;charset=UTF-8", "application/json; charset=UTF-8"],
            xml: ["application/xml", "application/xml;charset=utf-8", "application/xml; charset=utf-8", "application/xml;charset=UTF-8", "application/xml; charset=UTF-8"]
        }
    }
};

// Create Solr client, execute query and print number of documents in response.
solrSmartClient.createClient('my_solr_collection', options, function (err, solrClient) {
    if (err) {
        return console.log(err);
    }
    solrClient.search('q=*:*', function (err, obj) {
        if (err) {
            return console.log(err);
        }
        console.log('Number of documents found: %d', obj.response.numFound);
    })
});


lib/zookeeper-utils.js
var zkClient = require('node-zookeeper-client');

var logger = null;

function zkConnectAndExecute(dto, callback) {
    if (!dto) {
        return callback(new Error('Missing DTO'));
    }

    var client = zkClient.createClient(dto.zkConnectionString, dto.zkOptions);
    client.once('connected', function () {
        callback(null, client);
        client.close();
    });
    client.connect();
}

function zkExistsPath(dto, path, callback) {
    zkConnectAndExecute(dto, function (err, client) {
        if (err) {
            return callback(err);
        }
        client.exists(path, function (err, stat) {
            if (err) {
                return callback(err);
            }
            callback(null, stat !== null);
        });
    });
}

function zkCreatePathAndData(dto, path, data, callback) {
    zkConnectAndExecute(dto, function (err, client) {
        if (err) {
            return callback(err);
        }
        client.create(path, data, callback);
    });
}

function zkSetData(dto, path, data, callback) {
    zkConnectAndExecute(dto, function (err, client) {
        if (err) {
            return callback(err);
        }
        client.setData(path, data, callback);
    });
}

function zkGetData(dto, path, callback) {
    zkConnectAndExecute(dto, function (err, client) {
        if (err) {
            return callback(err);
        }
        client.getData(path, callback);
    });
}

function zkGetChildren(dto, path, callback) {
    zkConnectAndExecute(dto, function(err, client) {
        if (err) {
            return callback(err);
        }
        client.getChildren(path, callback);
    })
}

function setLogger(lgr) {
    logger = lgr;
}

exports.existsPath = zkExistsPath;
exports.createPathAndData = zkCreatePathAndData;
exports.setData = zkSetData;
exports.getData = zkGetData;
exports.getChildren = zkGetChildren;
exports.setLogger = setLogger;


examples/query_waterfall.js
// Load dependencies
var solrSmartClient = require('../lib/index.js'),
    async = require('async');

// Define options
options = {
    zkConnectionString: 'localhost:2181',
    zkLiveNodes: '/live_nodes',
    zkAliases: '/aliases.json',
    solrProtocol: 'http',
    solrCollectionsGetEndPoint: '/admin/collections?action=LIST&wt=json', // Supports XML and JSON writer types
    ssh: {},
    // Passed verbatim to node-zookeeper-client
    zk: {
        sessionTimeout: 3000,
        spinDelay : 1000,
        retries : 1
    },
    // Passed verbatim to node-rest-client
    rest: {
        requestConfig: {
            timeout: 3000
        },
        responseConfig: {
            timeout: 3000
        },
        mimetypes: {
            json: ["application/json", "application/json;charset=utf-8", "application/json; charset=utf-8", "application/json;charset=UTF-8", "application/json; charset=UTF-8"],
            xml: ["application/xml", "application/xml;charset=utf-8", "application/xml; charset=utf-8", "application/xml;charset=UTF-8", "application/xml; charset=UTF-8"]
        }
    }
};

// In waterfall approach
async.waterfall([
    function (callback) {
        // ..create Solr client,...
        solrSmartClient.createClient('my_solr_collection', options, callback);
    },
    function (solrClient, callback) {
        // ...execute query...
        solrClient.search('q=*:*', callback);
    }], function (err, obj) {
    // ...and print number of documents in response
    if (err) {
        return console.log(err);
    }
    console.log('Number of documents found: %d', obj.response.numFound);
});


lib/index.js
var SolrClient = require('solr-client'),
    RESTClient = require('node-rest-client').Client,
    async = require('async'),
    url = require('url'),
    _ = require('underscore'),
    zkUtils = require('./zookeeper-utils.js');

var DEFAULT_OPTIONS = {
    zkConnectionString: 'localhost:2181',
    zkLiveNodes: '/live_nodes',
    zkAliases: '/aliases.json',
    solrProtocol: 'http',
    solrCollectionsGetEndPoint: '/admin/collections?action=LIST&wt=json',
    ssh: {},
    // Passed verbatim to node-zookeeper-client
    zk: {
        sessionTimeout: 3000,
        spinDelay: 1000,
        retries: 1
    },
    // Passed verbatim to node-rest-client
    rest: {
        requestConfig: {
            timeout: 3000
        },
        responseConfig: {
            timeout: 3000
        },
        mimetypes: {
            json: ["application/json", "application/json;charset=utf-8", "application/json; charset=utf-8", "application/json;charset=UTF-8", "application/json; charset=UTF-8"],
            xml: ["application/xml", "application/xml;charset=utf-8", "application/xml; charset=utf-8", "application/xml;charset=UTF-8", "application/xml; charset=UTF-8"]
        }
    }
};

function createClient(solrCollection, options, callback) {
    options = _.defaults(_.clone(options), DEFAULT_OPTIONS);
    parseSshTunnels(options.ssh);

    async.waterfall([
        function (callback) {
            getZkInfo(options, callback);
        },
        function (zkInfo, callback) {
            getSolrCollections(zkInfo, options, callback);
        },
        function (solrBundles, callback) {
            getSolrInstance(solrBundles, solrCollection, callback);
        },
        function (solrInstance, callback) {
            getSolrParams(solrInstance, options, callback);
        }
    ], function (err, solrParams) {
        if (err) {
            return callback(err);
        }
        callback(null, SolrClient.createClient(solrParams.host, solrParams.port, solrParams.core, solrParams.path + '/' + solrCollection));
    });
}

function getSolrParams(solrInstance, options, callback) {
    solrInstance = tunnel(solrInstance, options.ssh.tunnels);
    var urlParams = url.parse(solrInstance);
    callback(null, {
        host: urlParams.hostname,
        port: urlParams.port,
        core: '',
        path: urlParams.path
    })
}

function getSolrInstance(solrBundles, solrCollection, callback) {
    async.filter(solrBundles, function (bundle, callback) {
        callback(bundle.collections.indexOf(solrCollection) > -1);
    }, function (filteredBundles) {
        async.map(filteredBundles, function (bundle, callback) {
            callback(null, bundle.instance);
        }, function (err, instances) {
            if (err) {
                return callback(err);
            }
            if (!instances.length) {
                return callback('Found no Solr instance hosting collection \'' + solrCollection + '\'');
            }
            var instance = instances[Math.floor(Math.random() * instances.length)];
            callback(null, instance);
        })
    });
}

function createSolrCollectionGETRequestTasks(zkInfo, options) {
    var restClient = new RESTClient(options.rest);
    return zkInfo.liveNodes.map(function (liveNode) {
        if (options.ssh && options.ssh.tunnels) {
            liveNode = tunnel(liveNode, options.ssh.tunnels);
        }
        var protocolLiveNode = options.solrProtocol + '://' + liveNode;
        return function (callback) {
            restClient.get(protocolLiveNode + options.solrCollectionsGetEndPoint, options.rest, function (data, response) {
                var instanceCollections;
                if (isXMLRESTResponse(response)) {
                    instanceCollections = data.response.arr[0].str;
                } else if (isJSONRESTResponse(response)) {
                    instanceCollections = data.collections;
                } else {
                    return callback('Unsupported content type \'' + response.headers['content-type'] + '\' in response of \'' + response.req.path + '\'');
                }
                var bundle = {
                    instance: protocolLiveNode,
                    collections: instanceCollections
                };
                callback(null, bundle);
            });
        };
    });
}

function getSolrCollections(zkInfo, options, callback) {
    var tasks = createSolrCollectionGETRequestTasks(zkInfo, options);
    async.parallel(tasks, function (err, bundles) {
        if (err) {
            return callback(err);
        }
        if (zkInfo.aliases) {
            bundles.forEach(function (bundle) {
                var collectionAliases = [];
                bundle.collections.forEach(function (collection) {
                    if (zkInfo.aliases[collection]) {
                        collectionAliases.push(zkInfo.aliases[collection]);
                    }
                });
                Array.prototype.push.apply(bundle.collections, collectionAliases);
            });
        }
        callback(null, bundles);
    });
}

function getZkInfo(options, callback) {
    var dto = {zkConnectionString: options.zkConnectionString, zkOptions: options.zk};
    async.parallel({
        liveNodes: function (callback) {
            zkUtils.getChildren(dto, options.zkLiveNodes, function (err, children /*, stats*/) {
                if (err) {
                    return callback(err);
                }
                if (children.length == 0) {
                    return callback('Found no live Solr nodes under path \'' + options.zkLiveNodes + '\' by connecting at \'' + options.zkConnectionString + '\'');
                }
                async.map(children, function (item, callback) {
                    var url = replaceAfterPort(item, /_/g, '/');
                    callback(null, url);
                }, callback);
            });
        },
        aliases: function (callback) {
            if (options.zkAliases) {
                zkUtils.getData(dto, options.zkAliases, function (err, data /*, stats*/) {
                    if (err) {
                        return callback(err);
                    }
                    var dataObj = data ? JSON.parse(data.toString()) : {};
                    if (dataObj.collection) {
                        var aliases = swap(dataObj.collection);
                    }
                    callback(null, aliases);
                });
            } else {
                callback();
            }
        }
    }, callback);
}

function isXMLRESTResponse(response) {
    return response.headers['content-type'].indexOf('application/xml') > -1;
}

function isJSONRESTResponse(response) {
    return response.headers['content-type'].indexOf('application/json') > -1;
}

function parseSshTunnels(options) {
    if (typeof options.tunnels === 'string') {
        var tunnels = {};
        var tunnelsArr = options.tunnels.split(',');
        for (var i = 0, tunnel; tunnel = tunnelsArr[i]; i++) {
            var portHostPort = tunnel.split(':');
            tunnels[portHostPort[1] + ':' + portHostPort[2]] = 'localhost:' + portHostPort[0];
        }
        options.tunnels = tunnels;
    }
}

function tunnel(instance, tunnels) {
    var pattern = /([^:]+:\d+)/;
    if (tunnels && pattern.test(instance)) {
        var hostPort = pattern.exec(instance)[1];
        if (tunnels.hasOwnProperty(hostPort)) {
            instance = instance.replace(hostPort, tunnels[hostPort]);
        }
    }
    return instance;
}

function replaceAfterPort(url, pattern, replacement) {
    var colonDigits = /:\d+/.exec(url)[0];
    var pathIndex = url.indexOf(colonDigits) + colonDigits.length;
    var hostPort = url.substring(0, pathIndex);
    var path = url.substring(pathIndex);
    return hostPort + path.replace(pattern, replacement);
}

function swap(obj) {
    var ret = {};
    for (var key in obj) {
        if (obj.hasOwnProperty(key)) {
            ret[obj[key]] = key;
        }
    }
    return ret;
}

exports.createClient = createClient;
